### DISCUSSION TITLE
Table selections boxes with custom body slot

### DISCUSSION THREAD

**Discussion by derkin123** [zauberzeug/nicegui#1192] [Q&A]
### Question

I want to create a table where each item in a column can be edited by the user via a popup edit, but I also want the user to be able to select rows using the selection='multiple' parameter of the table.

When you do this by manually specifying the body of the table it gets rid of the selection boxes. What do I need to add to the body slot definition to be able to also leverage the selection boxes?

An example of what I am trying to do:

```Python
from nicegui import ui

columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
]
rows = [
    {'id': 0,'name': 'Alice', 'age': 42},
    {'id': 1, 'name': 'Bob', 'age': 23},
]

with ui.table(columns, rows, row_key='id', selection='multiple').classes('w-full') as table:

    def TestPrint(msg):
        print(msg)

    table.add_slot('body', r'''
        <q-tr :props="props">
            <q-td key="name" :props="props">
                {{ props.row.name }}
                <q-popup-edit v-model="props.row.name" v-slot="scope" @save="(value, initialValue) => $parent.$emit('TestPrint', {id: props.row.id, name: props.row.name, new_name: value})">
                    <q-input v-model="scope.value" dense autofocus @keyup.enter="scope.set" />
                </q-popup-edit>
            </q-td>
            <q-td key="age" :props="props">
                {{ props.row.age }}
            </q-td>
        </q-tr>
    ''')
    table.on('TestPrint', TestPrint)

ui.run(dark=True, native=True)
```

**Comment by falkoschindler** [comment]
Hi @derkin123!

Quasar's QTable has a slot "body-cell-[name]" where "name" is a column name of your choice. This is useful for defining individual cells, without the need for writing the template for a whole row. So we don't have to deal with the checkbox:
```py
    table.add_slot('body-cell-name', r'''
        <q-td key="name" :props="props">
            {{ props.row.name }}
            <q-popup-edit v-model="props.row.name" v-slot="scope" @save="(value, initialValue) => $parent.$emit('TestPrint', {id: props.row.id, name: props.row.name, new_name: value})">
                <q-input v-model="scope.value" dense autofocus @keyup.enter="scope.set" />
            </q-popup-edit>
        </q-td>
    ''')
    table.add_slot('body-cell-age', r'''
        <q-td key="age" :props="props">
            {{ props.row.age }}
        </q-td>
    ''')
```

**Reply by derkin123** [reply]
Thank you, that was extremely helpful!

I have a slightly related follow up question. How would I best go about formatting a column to a currency when it is being specified in this manner?

My goal would be for this currency column to be along the lines of:

```Python
table.add_slot('body-cell-name', r'''
      <q-td key="currency" :props="props">
          {{ props.row.currency}}
          <q-popup-edit v-model="props.row.currency" v-slot="scope" @save="<some function>">
              <q-input v-model="scope.value" dense autofocus @keyup.enter="scope.set" />
          </q-popup-edit>
      </q-td>
  ''')
```
 I know that the column definition has a field format:

```Python
columns = [{'name': 'age', 'label': 'Age', 'field': 'age', ':format': 'value => value + " years" }]
```
but this doesn't seem to work for me when we define columns manually this way, I also don't know how I would force that to give commas in a currency or force only 2 decimal places (the 2 decimal places is my greatest point of emphasis in this formatting).

I've also looked into QInput and found it has [masks ](https://quasar.dev/vue-components/input#using-third-party-mask-processors) you can use. One specifically called out is v-money, but I don't know where I would go and specify either the moneyFormatForDirective or moneyFormatForComponent script blocks they use.

**Reply by falkoschindler** [reply]
I don't know. I would have to dig around myself for a while. Just a quick note: For your currency slot you should use the slot name "body-cell-currency" (if "currency" is the `field` of this column).

**Comment by derkin123** [comment]
@falkoschindler I've made some progress on this! I found that in the column format you can specify the value's format in the following way to get a monetary format in any locale:

```Python
':format': 'value => "$"+ parseFloat(value).toLocaleString("en-us", {minimumFractionDigits: 2, maximumFractionDigits: 2})', },
 ```

This method I've found is a valid solution if you only want a table in which you don't specify a body slot for a specific column containing the currency like this:
```Python
from nicegui import ui

columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'currency', 'label': 'Currency', 'field': 'currency', ':format': 'value => "$"+ parseFloat(value).toLocaleString("en-us", {minimumFractionDigits: 2, maximumFractionDigits: 2})', },
]
rows = [
    {'id': 0,'name': 'Alice', 'currency': 123456.789},
    {'id': 1, 'name': 'Bob', 'currency': 987654.321},
]
ui.table(columns, rows, row_key='id', selection='multiple').classes('w-full')

ui.run(dark=True, native=True)
```

If you want to specify the currency column to have an input then you can do it with a mask in the following manner:
```Python
from nicegui import ui

columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'currency', 'label': 'Currency', 'field': 'currency', ':format': 'value => "$"+ parseFloat(value).toLocaleString("en-us", {minimumFractionDigits: 2, maximumFractionDigits: 2})', },
]
rows = [
    {'id': 0,'name': 'Alice', 'currency': 123456.789},
    {'id': 1, 'name': 'Bob', 'currency': 987654.321},
]
with ui.table(columns, rows, row_key='id', selection='multiple').classes('w-full') as table:
    def TestPrint(msg):
        print(msg)

    table.add_slot('body-cell-currency', r'''
        <q-td key="currency" :props="props">
            ${{ props.row.currency }}
            <q-popup-edit v-model="props.row.currency" v-slot="scope" @save="(value, initialValue) => $parent.$emit('TestPrint', {id: props.row.id, amount: props.row.currency, new_amount: value})">
                <q-input filled v-model="scope.value" unmasked-value mask="###,###,###.##" reverse-fill-mask dense autofocus @keyup.enter="scope.set" />
            </q-popup-edit>
        </q-td>
    ''')
    table.on('TestPrint', TestPrint)

ui.run(dark=True, native=True)
```

There are a couple of issues with this second method:

1. With v-model="scope.value" + unmasked-value you will get no masking to appear, but if you remove unmasked-value then the back-end value will require parsing as it will still contain any commas in the mask. Is there a way to specify the v-model to contain the visual value and not the back-end value like they do here? https://quasar.dev/vue-components/input#example--unmasked-model
2. Default values don't seem to have the mask applied, so in the data array specified above the mask isn't applied until the user manually specifies a new value, any thoughts on how the could be enforced on default values? I couldn't find anything related to this on the documentation. 


I think part of the difficulties here comes from using an input to format the value of the table when its really only meant to format itself. If it was possible to pass along the format parameter in the column definition to the table when specifying a column manually then it would be much easier to format custom tables like this.

**Reply by falkoschindler** [reply]
Thanks for the update, @derkin123! Great to hear that you made progress on this subject. 🙂
It sounds like you're deep in Quasar land and I'm a bit lost. You might be able to reproduce the problem with a [standalone Quasar app](https://quasar.dev/start/umd). This might be a way to isolate your questions from NiceGUI and ask in a forum with more Quasar expertise.

**Reply by derkin123** [reply]
@falkoschindler Yes, I am very deep down the Quasar rabbit hole but I think I know what I am trying to do from a Quasar perspective and have questions on how to do this (if possible) with nicegui.

A simple code pen of what I want can be found [here](https://codepen.io/slanderkin/pen/Jjeeazz). Essentially I want to add a JS function that will allow me to format values in custom body-cell slots. This seems to be the way to format body-cell values since the format parameter in the column definition won't be used when defining body-cell slots.

In nicegui it would be something like this:
```Python
table.add_slot('body-cell-currency', r'''
  <q-td>
        {{formatCurrency(props.row.currency)}}
  </q-td>
''')
```

but having the function formatCurrency defined somehow like it is in the code pen above. I have tried using something along the lines of:
```Python
ui.add_body_html('''
    <script>
         function formatCurrency(val) {
            return "$"+ parseFloat(val).toLocaleString("en-us", {minimumFractionDigits: 2, maximumFractionDigits: 2});
         }
    </script>
''')
```

both with and without the function tag but this gives an error that the formatCurrency function can't be found. I saw a ui.run_javascript function but it doesn't seem to apply to this situation, or at least I don't know how I would make it work. So my question boils down to is there a way with nicegui to add a JS function so that the html within my slot definitions can reference/use it?

**Reply by falkoschindler** [reply]
You should be able to define JS functions using `app.add_head_html` or `app.add_body_html`. They should be available for formatting table cells. Haven't tried it though.
